src/main/host/host.rs

```diff
@@ -412,8 +412,9 @@ impl Host {
         crate::core::logger::log_wrapper::c_to_rust_log_level(level).map(|l| l.to_level_filter())
     }
 
-    pub fn upstream_router(&self) -> *mut Router {
-        unsafe { cshadow::hostc_getUpstreamRouter(self.chost()) }
+    pub fn with_upstream_router_mut<Res>(&self, f: impl FnOnce(&mut Router) -> Res) -> Res {
+        let router = unsafe { &mut *cshadow::hostc_getUpstreamRouter(self.chost()) };
+        f(router)
     }
```

Comment:
This doesn't seem safe? It creates a `&mut` from ` &` without any runtime checks?

Reply:
I see that a `RefCell` was added in a later commit.